cuML/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,93 @@
+
+# Logging
+Add once https://github.com/rapidsai/cuml/issues/100 is addressed.
+
+# Device and Host memory allocations
+To enable `libcuml.so` users to control how memory for temporary data is allocated device memory should only be allocated via the allocator provided:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getDeviceAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    handle->getDeviceAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+the same rule applies to larger amounts of host heap memory:
+```cpp
+template<typename T>
+void foo(ML::cumlHandle* handle, cudaStream_t stream, ... )
+{
+    T* temp_h = handle->getHostAllocator()->allocate(n*sizeof(T), stream);
+    ...
+    handle->getHostAllocator()->deallocate(temp_h, n*sizeof(T), stream);
+}
+```
+Small host memory heap allocations, e.g. as internally done by STL containers, are fine, e.g. an `std::vector` managing only a handful of integers.
+Both the Host and the Device Allocators might allow asynchronous stream ordered allocation and deallocation. This can provide significant performance benefits so a stream always needs to be specified when allocating or deallocating (see [Asynchronous operations and stream ordering](# Asynchronous operations and stream ordering)).
+There are two simple container classes compatible with the `ML::cumlHandle` allocator interface `ML::device_buffer` available in ` cuML/src/common/device_buffer.hpp` and `ML::host_buffer` available in ` cuML/src/common/host_buffer.hpp`. These allow to follow the [RAII idiom](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) to avoid resources leaks and enable exception safe code. These containers also allow asynchronous allocation and deallocation using the `resize` and `release` member functions:
+```cpp
+template<typename T>
+void foo( ML::cumlHandle* handle, .., cudaStream_t stream )
+{
+    ...
+    ML::device_buffer<T> temp( handle->getDeviceAllocator(), 0 )
+    
+    temp.resize(n, stream);
+    kernelA<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    kernelB<<<grid, block, 0, stream>>>(..., temp.data(), ...);
+    temp.release(stream);
+}
+```
+## Using thrust
+To ensure that thrust algorithms allocate temporary memory via the provided device memory allocator the `ML::thrustAllocatorAdapter` available in `allocatorAdapter.hpp` should be used with the `thrust::cuda::par` execution policy:
+```cpp
+void foo( ML::cumlHandle* handle, .. )
+{
+    ML::thrustAllocatorAdapter alloc( handle->getDeviceAllocator(), handle->getStream() );
+    auto execution_policy = thrust::cuda::par(alloc).on(handle->getStream());
```

Comment:
Thanks for pointing this out. This is a great idea. I just tried it but to my understanding the implementation of `exec_policy` requires C++14 and I could not find a elegant way to express it with C++11 (which cuML is currently at). So I an implementation of `exec_policy` only conditionally and updated the developer guide mentioning the new possibilities with C++14. Let me know if I am missing something and we can make it work with C++11.

Reply:
C++14 is only required for the `auto` return type. 

It's possible to support C++11 using trailing return type: https://www.ibm.com/support/knowledgecenter/en/SSLTBW_2.3.0/com.ibm.zos.v2r3.cbclx01/trailing_return.htm

It would just be a little more verbose, something like...
```
using T = decltype(thrust::cuda::par(rmm_allocator<char>{}));
inline exec_policy(cudaStream_t stream = 0) -> std::unique_ptr<T, std::function<void(T*)>
```

Reply:
Thanks @jrhemstad I could make it work with C++11 and just pushed the related updates.